### ISSUE TITLE
PSA: Downgrade linux-firmware and kernel if models start to behave strangely

### ISSUE THREAD

**Issue by rabejens** [noamsto/nix-amd-ai#105]
Not an issue, but something I found and I thought was worth sharing. Maybe this should be added to the docs.

I am on nixpkgs-unstable, and for some time now, my models started to behave strangely. Qwen 3.8 started running into tool-calling loop (calling a tool failed, it noticed it, said it wanted to try another one, then called the same tool over and over) and Qwen 3.6 stopped in the middle of doing a task in an agent harness.

I pinned this down to Kernel 7.2.x and linux-firmware 20260810 - as around the time when that was updated, the problems began.

The last commit that still had linux-firmware 20260622 (the good one) was [d7df67e](https://github.com/NixOS/nixpkgs/commit/d7df67e2ba27de7e4ecd05bd15824b7c09d11831). Unfortunately this firmware is not compatible with the 7.2.x kernels and also not with the normal `pkgs.linuxPackages`, so you have to override both.

Add to your flake inputs:

```nix
nixpkgs-firmware.url = "github:nixos/nixpkgs?rev=d7df67e2ba27de7e4ecd05bd15824b7c09d11831";
```

and the following override:

```nix
final: prev:
let
  firmware = inputs.nixpkgs-firmware.legacyPackages.${prev.stdenv.hostPlatform.system};
in
{
  linux-firmware = firmware.linux-firmware;
  linuxPackages_latest = firmware.linuxPackages_latest;
  # linuxPackages = firmware.linuxPackages; # NOT RECOMMENDED - SEE BELOW
}
```

If you use other kernels, override them, too.

This will pin the kernel at 7.1.8 when using latest, and the firmware at 20260622 and models return to normal behaviour.

It happened on a Minisforum S1-MAX (Strix Halo). It might differ from machine to machine because I haven't seen this phenomenon on a Framework Desktop at work which has the same Strix Halo.

**Followup** - You should use and override `linuxPackages_latest`. If you override the default `linuxPackages`, it will build a huge part of the system from source at that point in time.

**Comment by noamsto** [comment]
Thanks for writing this up — it landed at a useful moment. Our Halo box is now running, and it is on **exactly the pair you flagged**:

```
kernel          7.2.2
linux-firmware  20260810
```

So this is reproducible territory for us, and it has already changed our plans: `docs/halo-bringup-checklist.md` puts the expensive ROCm-vs-Vulkan long-context A/B (#61) last, on the logic that cheap phases exist to make those numbers trustworthy. Benchmarking on a firmware a Halo owner reports as behaviourally broken would produce numbers we would have to throw away — so this issue now gates that work rather than sitting behind it.

What I can confirm from our host so far, all Phase 0 material and none of it contradicting you:

| | |
|---|---|
| CPU / iGPU | Ryzen AI MAX+ 395, `gfx1151` (correctly identified, not `gfx1100`) |
| NPU | rev `0x11` → npu5, fw `1.1.2.65` |
| kernel / firmware | **7.2.2 / 20260810** |

What I have **not** done is reproduce the misbehaviour itself, or A/B it against the pinned 20260622 firmware — that needs the overlay and a reboot. So I am not yet in a position to confirm or dispute the diagnosis, only to say we are on the same configuration.

Two things that would sharpen it, if you have them handy:

1. Whether it reproduces on a **fixed seed / temperature 0** prompt, or only in longer agent-harness runs. A deterministic repro would separate "firmware changed model behaviour" from "longer sessions hit it more often".
2. Whether the Framework Desktop you mention (same Strix Halo, no symptom) was on the same kernel *and* the same firmware, or differed on one of them. Right now that machine is the strongest evidence for and against a firmware cause, depending on which.

Also noting your point about the docs: nothing in `README.md` or `docs/` currently mentions `linux-firmware` at all. I would rather write that up once we have the A/B than assert a cause we have not measured — this repo has a rule about hardware-gated claims and this is squarely one. Leaving the issue open as the tracking item for both.

Your followup about overriding `linuxPackages_latest` rather than `linuxPackages` is a genuinely valuable detail and would have cost someone an afternoon of source builds.

**Comment by noamsto** [comment]
Thanks for this — the report was right that something is badly wrong, and it's what prompted a full investigation on a Halo box running the exact pair you named (kernel 7.2.2, `linux-firmware` 20260810). Writeup with method and raw numbers: [`docs/rocm-gfx1151-numerics.md`](https://github.com/noamsto/nix-amd-ai/blob/main/docs/rocm-gfx1151-numerics.md).

**Confirmed: models really do produce garbage.** Measured with `llama-perplexity` over a fixed corpus (deterministic — repeat runs are byte-identical):

| Backend | PPL |
| --- | ---: |
| CPU | 6.8056 |
| Vulkan | 6.8067 |
| **ROCm** | **1334.00** |

A perplexity of 1334 against 6.81 is near-random token selection. That matches your symptoms exactly — a tool called over and over, a task abandoned midway, is what broken logits look like from inside an agent harness.

**But the cause looks like the ROCm backend, not the kernel or firmware.** Firmware and kernel are shared by all three backends, and two of the three are numerically perfect on the firmware you flagged. A shared cause can't produce a backend-specific failure. I also tested the one firmware-carried path Vulkan never touches (`HSA_ENABLE_SDMA=0`) — no change.

The strongest evidence: AMD's own prebuilt gfx1151 binary ([`lemonade-sdk/llamacpp-rocm`](https://github.com/lemonade-sdk/llamacpp-rocm) b1325), which bundles **ROCm 10.1.0a20260822** and a different llama.cpp commit, reproduces it at the same magnitude (1334.57). Two ROCm major versions, two llama.cpp commits, a vendor-built binary — same failure.

**So the practical fix is a one-line recipe change, not a system rebuild:**

```
llamacpp:rocm  ->  llamacpp:vulkan
```

Vulkan is correct here and, on the benchmarks in this repo's README, already faster on decode. Worth trying before the firmware/kernel pin — if your models behave again, you've saved yourself a large rebuild and an override to maintain.

If Vulkan *doesn't* fix it for you, that's genuinely useful information and worth saying here, because it would mean there's a second problem the firmware pin was addressing.

One caveat in both directions: everything above is from one Halo host with one model, so it may not transfer to your Minisforum S1-MAX. And this doesn't prove firmware is irrelevant — only that it can't be the whole story, since Vulkan and CPU are fine on it. The A/B against 20260622 hasn't been run.

Also worth noting for anyone landing here: this looks like a live upstream bug, not something already fixed. Current llama.cpp master still reproduces it, and the per-layer half is *worse* there than in the version we ship — so bumping llama.cpp is not the answer either.

**Comment by rabejens** [comment]
Hello,

thanks for the investigation. I am not going to Vulkan because I also use this machine as a workstation, not as a pure AI server. When using Vulkan, both VRAM and GTT are filled. And since I am running my GUI on X11 instead of Wayland, as I also use it via RDP and Wayland does not support RDP, I get random lockups to a point where even SysRQ REISUB does not bring the computer back to life when VRAM is filled to the brim. Using ROCm might be a bit slower, but this only uses GTT and keeps VRAM free.

I did notice an improvement of output quality after the downgrade. But it can also be the case that by chance, that operation (which updated other flake inputs at the same time) at the same time pulled in a gfx1151 binary that fixed the problem at the same time. This might also explain why I haven't noticed these problems at the Framework Desktop at work despite it running a 7.2.x kernel and Firmware 20260810 - they might have been at slightly different versions of the stack.

But I can confirm my models DO run fine with the current versions on Kernel 7.1.8 with firmware 20260622.

**Comment by noamsto** [comment]
Follow-up: **there's a real fix, and it's not the firmware.** My earlier comment said no fix existed — that was wrong, and I should have searched upstream harder before saying it.

[ggml-org/llama.cpp#28211](https://github.com/ggml-org/llama.cpp/issues/28211) reports the same thing on gfx1151 (wrong logits, no crash, Vulkan fine), and @Error00101 posted a patch there — `865374bb`, *"hip: disable direct host access on gfx1151"*, eight lines in `ggml-cuda.cu`.

**Root cause:** gfx1151 reports as an integrated GPU, so ggml let tensors live in *host* memory for the GPU to read directly. On this chip that hands back wrong data, before any arithmetic happens. That explains why it looked so strange from the outside — the corruption was independent of quantization, of the matmul path, and of ROCm version.

Verified on the Halo box:

| `-ngl` | before | after |
| --- | ---: | ---: |
| 0 | 6.8024 | 6.8114 |
| 32 | 15.6257 | **6.8114** |
| 99 | 1326.89 | **6.8280** |

ROCm now matches CPU (6.81) and Vulkan (6.81) at every offload depth.

This flake carries the patch as of #128, so `llamacpp:rocm` works again on gfx1151 — **no firmware or kernel downgrade needed**. If you're not on this flake, the patch applies cleanly to current llama.cpp master, or you can wait for it to land upstream.

Worth noting for your own setup: `#28211` also has a report that lemonade's `llama-rocm` **b1291** was the last known-good prebuilt. I tested **b1325** and it reproduces the bug, so if you use those binaries, that's a version worth avoiding.

Thanks for filing this — the symptom was real and the report is what started the investigation. The firmware correlation looks like coincidence, but it was a reasonable read given what you had.

**Comment by rabejens** [comment]
I read through the original bug description and it still baffles me how the kernel/firmware downgrade worked as a temporary workaround for me. I haven't looked into the dependencies of the nixpkgs too thoroughly, but might it be that this downgrade also pulled in some dependencies that didn't exhibit the bug and that was the original reason why the downgrade fixed the problem for me?

Anyway, thank you for quickly addressing that. I haven't yet had time to try it myself but I hop I get to it tomorrow.
